experimental/dds/tree2/src/domains/schemaBuilder.ts

```diff
@@ -35,17 +39,155 @@ import {
+	private readonly structuralTypes: Map<string, TreeSchema> = new Map();
+
 	public constructor(options: SchemaBuilderOptions<TScope>) {
 		super(FieldKinds.required, {
 			...options,
 			libraries: [...(options.libraries ?? []), leaf.library],
 		});
 	}
 
+	/**
+	 * Define (and add to this library if not already present) a structurally typed {@link TreeSchema} for a {@link FieldNode} of a {@link Sequence}.
+	 *
+	 * @remarks
+	 * The {@link TreeSchemaIdentifier} for this List is defined as a function of the provided types.
+	 * It is still scoped to this SchemaBuilder, but multiple calls with the same arguments will return the same schema object, providing somewhat structural typing.
+	 * This does not support recursive types.
+	 *
+	 * If using these structurally named lists, other types in this schema builder should avoid names of the form `List<${string}>`.
+	 *
+	 * @privateRemarks
+	 * The name produced at the type level here is not as specific as it could be, however doing type level sorting and escaping is a real mess.
+	 * There are cases where not having this full type provided will be less than ideal since TypeScript's structural types.
+	 * For example attempts to narrow unions of structural lists by name won't work.
+	 * Planned future changes to move to a class based schema system as well as factor function based node construction should mostly avoid these issues,
+	 * though there may still be some problematic cases even after that work is done.
+	 */
+	public list<const T extends TreeSchema | Any | readonly TreeSchema[]>(
+		allowedTypes: T,
+	): TreeSchema<
+		`${TScope}.List<${string}>`,
+		{
+			structFields: {
+				[""]: FieldSchema<typeof FieldKinds.sequence, NormalizeAllowedTypes<T>>;
+			};
+		}
+	>;
+
+	/**
+	 * Define (and add to this library) a {@link TreeSchema} for a {@link FieldNode} of a {@link Sequence}.
+	 *
+	 * The name must be unique among all TreeSchema in the the document schema.
+	 */
+	public list<Name extends TName, const T extends ImplicitAllowedTypes>(
+		name: Name,
+		allowedTypes: T,
+	): TreeSchema<
+		`${TScope}.${Name}`,
+		{
+			structFields: {
+				[""]: FieldSchema<typeof FieldKinds.sequence, NormalizeAllowedTypes<T>>;
+			};
+		}
+	>;
+
+	public list<const T extends ImplicitAllowedTypes>(
+		nameOrAllowedTypes: TName | ((T & TreeSchema) | Any | readonly TreeSchema[]),
+		allowedTypes?: T,
+	): TreeSchema<
+		`${TScope}.${string}`,
+		{
+			structFields: {
+				[""]: FieldSchema<typeof FieldKinds.sequence, NormalizeAllowedTypes<T>>;
+			};
+		}
+	> {
+		if (allowedTypes === undefined) {
+			let inner: string;
+			const types = nameOrAllowedTypes as (T & TreeSchema) | Any | readonly TreeSchema[];
+			if (types === Any) {
+				inner = "Any";
+			} else if (types instanceof TreeSchema) {
+				// Call back into this with a array to ensure version with array gets named identical to version with single item.
+				return this.list([types]) as TreeSchema<
```

Comment:
why is this cast needed? if it is, can we clean it up using ReturnType<> magic?

Reply:
This is needed because without it you get a big untellable type error since TypeScript can tell they are equivalent.

As for ways to clean it up, I'm open to suggestions, but I'm not aware of a good one. If we were not using overloads here, there might be a way to use ReturnType and typeof that might be cleaner, but as is, I don't see one.